docs/projects/2024/architecture-transition/record-summaries.md

```diff
@@ -0,0 +1,79 @@
+
+## New record summaries
+
+_This section describes how we intend for record summaries to work in Mathesar's beta release._
+
+Without any user configuration, records will be summarized by their first _text-like_ column. This is a slight change. For example, if the PK is text, then we use that. If the PK is numeric and the first non-pk column is a numeric FK column, then we keep looking to see if we can find a text column.
+
+Users will still be able to customize the template, but the template DSL will have slightly new syntax. Column references within curly braces will need to be quoted in the same manner as in SQL. So `{first_name} {last_name}` will work if the columns are `first_name` and `last_name`. But if the columns are `First Name` and `Last Name`, then the template will need to be `{"First Name"} {"Last Name"}`. The same quoting and escaping rules apply as with SQL. Any references which fail to parse correctly are interpreted as string literals to be included in the template verbatim. Users should rarely be interacting with the template DSL manually because Mathesar has (and will continue to have) a button to insert a column into the template.
+
+Record summaries will no longer use the transitive templates of other tables referenced via foreign key columns. So from the Books table, putting `{"Author"}` in a template will not show the _summarized_ author (as before). Instead, it will only show the PK value of the author. To show fields in the author's table, we use dot notation within the curly braces.
+
+From the same example above, to configure the record summary for the Books table so that books are summarized like "Blink, by Malcom Gladwell", we would not need to configure the record summary template for the Authors table. We would configure the record summary template for the Books table as `{"Title"}, by {"Author"."First Name"} {"Author"."Last Name"}`.
+
+The UX for inserting things like `{"Author"."First Name"}` into the template is still TBD, but we expect it to function akin to way we add columns to explorations.
+
+From the user's perspective, the change in the transitive behavior has some advantages and some disadvantages. It's slightly more flexible — but also slightly more tedious. Most importantly though: this modification to the transitive behavior helps us achieve our performance goals.
+
+The record summary template will still be persisted in metadata associated with a table. But instead of persisting it as a string, it will be persisted as a parsed syntax tree. For the example above, that tree looks like this:
+
+```json
+[
+  [2],
+  ", by ",
+  [10, 2],
+  " ",
+  [10, 3]
+]
+```
+
+That means: first include a reference to the column in the base table having attnum 2; then include a string literal, then include a reference to a column that can be found by joining on the FK at column 10 and finding column with attnum 2 in the joined table... and so on.
```

Comment:
I think we should reuse the `fkey_path` and/or `jp_path` structures we already have defined for holding this info. Out of curiosity , are you planning to be able to go multiple joins deep? In that case I _really_ think we should use the more involved structures. They're more flexible in the long-run since they can handle arbitrary joins (at least in the case of `jp_path`). It also lets us reuse some machinery we've already build for getting the needed info (see `1_msar_joinable_tables.sql` for example).